element-connector/src/main/java/org/eclipse/californium/elements/auth/PreSharedKeyIdentity.java

```diff
@@ -21,32 +21,82 @@
  * A principal representing an authenticated peer's identity as used in a
  * <em>pre-shared key</em> handshake.
  */
-public class PreSharedKeyIdentity implements Principal {
+public final class PreSharedKeyIdentity implements Principal {
 
+	private final String virtualHost;
 	private final String identity;
+	private final int hash;
+	private final String name;
 
 	/**
-	 * Creates a new instance for a given identity.
+	 * Creates a new instance for an identity.
 	 * 
 	 * @param identity the identity
 	 * @throws NullPointerException if the identity is <code>null</code>
 	 */
 	public PreSharedKeyIdentity(String identity) {
+		this(null, identity);
+	}
+
+	/**
+	 * Creates a new instance for an identity scoped to a virtual host.
+	 * 
+	 * @param virtualHost the virtual host name that the identity is scoped to.
+	 * @param identity the identity.
+	 * @throws NullPointerException if the identity is <code>null</code>
+	 */
+	public PreSharedKeyIdentity(String virtualHost, String identity) {
 		if (identity == null) {
 			throw new NullPointerException("Identity must not be null");
 		} else {
 			this.identity = identity;
+			this.virtualHost = virtualHost;
+
+			final int prime = 31;
+			int result = 1;
+			result = prime * result + ((identity == null) ? 0 : identity.hashCode());
+			result = prime * result + ((virtualHost == null) ? 0 : virtualHost.hashCode());
+			this.hash = result;
+
+			StringBuilder b = new StringBuilder();
+			if (virtualHost != null) {
+				b.append(virtualHost).append(":");
+			}
+			b.append(identity);
+			this.name = b.toString();
```

Comment:
???

Reply:
???

Reply:
> final int prime = 31;
> int result = 1;
> result = prime * result + ((identity == null) ? 0 : identity.hashCode());
> result = prime * result + ((virtualHost == null) ? 0 : virtualHost.hashCode());
> this.hash = result;

calculates a hash from `identity` and `virtualHost`, which functionality seems to be the same as `name.hashCode()` (the functionality not the value!). So I simply would rely on that `String.hashCode()` implementation, I can't see a need to precalculate a hash in that case.

Reply:
Oh, I see, you want to use the hash of the *name* property instead of calculating the hash from the identity and virtual host properties. That makes sense. :+1:

Reply:
taking another look, I now feel that comparing the two properties individually in equals() and computing a hash based on the two properties is more appropriate. Otherwise, the following two PreSharedKeyIdenties would be considered equal:
```
PreSharedKeyIdentity idOne = new PreSharedKeyIdentity("iot.eclipse", "org:device-1");
PreSharedKeyIdentity idTwo = new PreSharedKeyIdentity("iot.eclipse:org", "device-1");
```

Reply:
Are you sure, that a DNS hostname is allowed to contain ":org"?
FMPOV either special characters are reserved as separator and using them within the ids is a violation and reported as error, or, a "character escape rule" must be defined (e.g. if the ":"  should be part of the hostname, it must be encoded using %3A and it should not be de-escaped when using for the PSK name..

Reply:
> Are you sure, that a DNS hostname is allowed to contain ":org"?

That's a very good point :-) It isn't. I will add a check in both ServerName as well as PreSharedKeyIdentity to reject illegal DNS hostnames.